Connected peers exceed the upper bound peer limit

Some hoodi nodes (bn-01 and bn-02) started having more peers (190+) than the `p2p-peers-upper-bound` config parameters (set to 150).
This has been observed with he 25.5.0 teku version.
Logs examples:
- On bn-01 observed since June 15th

<img width="1308" alt="Image" src="https://github.com/user-attachments/assets/ebb3e650-4527-455c-b70e-4c5d5e8286ae" />

- On bn-02 observed since June 18th

<img width="1308" alt="Image" src="https://github.com/user-attachments/assets/9a9953c5-46be-4cd9-a7cd-55356594ee13" />

Our peers selection strategy tries to connect to a maximum number of peers (both inbound and outbound) and then disconnects some randomly selected ones in order to keep the peers count under the defined upper bound limit.
After debugging, the issue seems to be caused by peers disconnection issues where we end up keeping some peers that are supposed to be disconnected. The peers disconnection fails.

<img width="943" alt="Image" src="https://github.com/user-attachments/assets/b3498aa8-ee12-4d86-8159-fe74b8495825" />

The stack trace:

```
 java.util.concurrent.CompletionException: tech.pegasys.teku.networking.eth2.rpc.core.RpcException$ExtraDataAppendedException: [Code 1] Extra data appended to end of message:  extra data: ByteBuf{UnpooledSlicedByteBuf(ridx: 0, widx: 18, cap: 18/18, unwrapped: UnpooledHeapByteBuf(ridx: 0, widx: 37, cap: 37/37)), content: 0x0008ff060000734e61507059010c00000175}
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368)
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377)
	at java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at java.base/java.util.Optional.ifPresentOrElse(Optional.java:196)
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2RpcResponseHandler.onCompleted(Eth2RpcResponseHandler.java:118)
	at tech.pegasys.teku.networking.p2p.rpc.RpcResponseHandler.onCompleted(RpcResponseHandler.java:28)
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.lambda$abortRequest$5(Eth2OutgoingRequestHandler.java:247)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$always$29(SafeFuture.java:378)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$finish$33(SafeFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	at java.base/java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:950)
	at java.base/java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2372)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.handle(SafeFuture.java:613)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.finish(SafeFuture.java:391)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.always(SafeFuture.java:378)
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.abortRequest(Eth2OutgoingRequestHandler.java:247)
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.abortRequest(Eth2OutgoingRequestHandler.java:230)
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.processData(Eth2OutgoingRequestHandler.java:139)
	at tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.lambda$channelRead0$0(RpcHandler.java:217)
	at java.base/java.util.Optional.ifPresentOrElse(Optional.java:196)
	at tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.runHandler(RpcHandler.java:296)
	at tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.channelRead0(RpcHandler.java:217)
	at tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.channelRead0(RpcHandler.java:186)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1429)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
	at io.libp2p.etc.util.netty.mux.AbstractMuxHandler.childRead(AbstractMuxHandler.kt:71)
	at io.libp2p.mux.mplex.MplexHandler.channelRead(MplexHandler.kt:32)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:359)
	at tech.pegasys.teku.networking.p2p.libp2p.MuxFirewall$MuxFirewallHandler.channelRead(MuxFirewall.java:118)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:107)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:120)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:333)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:455)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1429)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:167)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.handle(AbstractNioChannel.java:445)
	at io.netty.channel.nio.NioIoHandler$DefaultNioRegistration.handle(NioIoHandler.java:383)
	at io.netty.channel.nio.NioIoHandler.processSelectedKey(NioIoHandler.java:577)
	at io.netty.channel.nio.NioIoHandler.processSelectedKeysOptimized(NioIoHandler.java:552)
	at io.netty.channel.nio.NioIoHandler.processSelectedKeys(NioIoHandler.java:493)
	at io.netty.channel.nio.NioIoHandler.run(NioIoHandler.java:470)
	at io.netty.channel.SingleThreadIoEventLoop.runIo(SingleThreadIoEventLoop.java:204)
	at io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:175)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1073)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: tech.pegasys.teku.networking.eth2.rpc.core.RpcException$ExtraDataAppendedException: [Code 1] Extra data appended to end of message:  extra data: ByteBuf{UnpooledSlicedByteBuf(ridx: 0, widx: 18, cap: 18/18, unwrapped: UnpooledHeapByteBuf(ridx: 0, widx: 37, cap: 37/37)), content: 0x0008ff060000734e61507059010c00000175}
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.processData(Eth2OutgoingRequestHandler.java:114)
	... 43 more
```

--- a/networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeer.java
+++ b/networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeer.java
@@ -142,18 +142,26 @@ public class LibP2PPeer implements Peer {
   public void disconnectImmediately(
       final Optional<DisconnectReason> reason, final boolean locallyInitiated) {
     if (connected.getAndSet(false)) {
-      disconnectReason = reason;
-      disconnectLocallyInitiated = locallyInitiated;
-      SafeFuture.of(connection.close())
-          .finish(
-              () ->
-                  LOG.trace(
-                      "Disconnected forcibly {} because {} from {}",
-                      locallyInitiated ? "locally" : "remotely",
-                      reason,
-                      connection.remoteAddress()),
-              error -> LOG.warn("Failed to disconnect from peer {}", getId(), error));
+      internalDisconnectImmediately(reason, locallyInitiated);
     }
+  }
+
+  private void internalDisconnectImmediately(
+      final Optional<DisconnectReason> reason, final boolean locallyInitiated) {
+    disconnectReason = reason;
+    disconnectLocallyInitiated = locallyInitiated;
+    SafeFuture.of(connection.close())
+        .finish(
+            () ->
+                LOG.trace(
+                    "Disconnected forcibly {} because {} from {}",
+                    locallyInitiated ? "locally" : "remotely",
+                    reason,
+                    connection.remoteAddress()),
+            error -> {
+              connected.set(true);
+              LOG.warn("Failed to disconnect from peer {}", getId(), error);
+            });
   }
 
   private SafeFuture<Optional<String>> getAgentVersionFromIdentity() {
@@ -180,7 +188,6 @@ public class LibP2PPeer implements Peer {
   public SafeFuture<Void> disconnectCleanly(final DisconnectReason reason) {
     if (connected.getAndSet(false)) {
       LOG.trace("Disconnecting cleanly because {} from {}", reason, connection.remoteAddress());
-      connected.set(false);
       disconnectReason = Optional.of(reason);
       disconnectLocallyInitiated = true;
       return disconnectRequestHandler
@@ -190,7 +197,7 @@ public class LibP2PPeer implements Peer {
                 if (error != null) {
                   LOG.debug("Failed to disconnect from " + getId() + " cleanly.", error);
                 }
-                disconnectImmediately(Optional.of(reason), true);
+                internalDisconnectImmediately(Optional.of(reason), true);
                 return null;
               });
     } else {

--- a/networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeer.java
+++ b/networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeer.java
@@ -188,14 +188,12 @@ public class LibP2PPeer implements Peer {
   public SafeFuture<Void> disconnectCleanly(final DisconnectReason reason) {
     if (connected.getAndSet(false)) {
       LOG.trace("Disconnecting cleanly because {} from {}", reason, connection.remoteAddress());
-      disconnectReason = Optional.of(reason);
-      disconnectLocallyInitiated = true;
       return disconnectRequestHandler
           .requestDisconnect(reason)
           .handle(
               (__, error) -> {
                 if (error != null) {
-                  LOG.debug("Failed to disconnect from " + getId() + " cleanly.", error);
+                  LOG.debug("Failed to disconnect from {} cleanly.", getId(), error);
                 }
                 internalDisconnectImmediately(Optional.of(reason), true);
                 return null;

--- a/networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeer.java
+++ b/networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeer.java
@@ -158,10 +158,7 @@ public class LibP2PPeer implements Peer {
                     locallyInitiated ? "locally" : "remotely",
                     reason,
                     connection.remoteAddress()),
-            error -> {
-              connected.set(true);
-              LOG.warn("Failed to disconnect from peer {}", getId(), error);
-            });
+            error -> LOG.warn("Failed to disconnect from peer {}", getId(), error));
   }
 
   private SafeFuture<Optional<String>> getAgentVersionFromIdentity() {

--- a/networking/p2p/src/test/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeerTest.java
+++ b/networking/p2p/src/test/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeerTest.java
@@ -15,16 +15,24 @@ package tech.pegasys.teku.networking.p2p.libp2p;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import io.libp2p.core.Connection;
 import io.libp2p.core.security.SecureChannel.Session;
 import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.IntStream;
+import kotlin.Unit;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
 import tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler;
+import tech.pegasys.teku.networking.p2p.peer.DisconnectReason;
 import tech.pegasys.teku.networking.p2p.reputation.ReputationManager;
 import tech.pegasys.teku.networking.p2p.rpc.RpcMethod;
 import tech.pegasys.teku.networking.p2p.rpc.RpcRequestHandler;
@@ -47,12 +55,14 @@ public class LibP2PPeerTest {
 
   private LibP2PPeer libP2PPeer;
 
+  private final SafeFuture<Unit> closeFuture = new SafeFuture<>();
+
   @BeforeEach
   public void init() {
     when(rpcHandler.getRpcMethod()).thenReturn(rpcMethod);
     final Session secureSession = mock(Session.class);
     when(connection.secureSession()).thenReturn(secureSession);
-    when(connection.closeFuture()).thenReturn(new SafeFuture<>());
+    when(connection.closeFuture()).thenReturn(closeFuture);
     libP2PPeer =
         new LibP2PPeer(connection, List.of(rpcHandler), ReputationManager.NOOP, peer -> 0.0);
   }
@@ -88,5 +98,35 @@ public class LibP2PPeerTest {
 
     // throttled request should have completed now
     assertThat(throttledRequest).isDone();
+  }
+
+  @Test
+  public void disconnectCleanly_shouldCloseConnectionOnlyOnce() {
+    final AtomicReference<Optional<DisconnectReason>> disconnectionReason = new AtomicReference<>();
+    final AtomicBoolean disconnectionLocallyInitiated = new AtomicBoolean();
+    final AtomicInteger disconnectionCount = new AtomicInteger(0);
+
+    libP2PPeer.subscribeDisconnect(
+        (reason, locallyInitiated) -> {
+          disconnectionReason.set(reason);
+          disconnectionLocallyInitiated.set(locallyInitiated);
+          disconnectionCount.addAndGet(1);
+        });
+
+    libP2PPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK);
+    verify(connection).close();
+
+    libP2PPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK);
+    libP2PPeer.disconnectImmediately(Optional.of(DisconnectReason.REMOTE_FAULT), false);
+
+    verify(connection, times(1)).close();
+
+    assertThat(disconnectionCount.get()).isEqualTo(0);
+
+    closeFuture.complete(null);
+
+    assertThat(disconnectionReason.get()).contains(DisconnectReason.IRRELEVANT_NETWORK);
+    assertThat(disconnectionLocallyInitiated.get()).isTrue();
+    assertThat(disconnectionCount.get()).isEqualTo(1);
   }
 }

--- a/networking/p2p/src/test/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeerTest.java
+++ b/networking/p2p/src/test/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeerTest.java
@@ -113,7 +113,7 @@ public class LibP2PPeerTest {
           disconnectionCount.addAndGet(1);
         });
 
-    libP2PPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK);
+    libP2PPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK).finish(__ -> {});
     verify(connection).close();
 
     libP2PPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK);

--- a/networking/p2p/src/test/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeerTest.java
+++ b/networking/p2p/src/test/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeerTest.java
@@ -116,7 +116,7 @@ public class LibP2PPeerTest {
     libP2PPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK).finish(__ -> {});
     verify(connection).close();
 
-    libP2PPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK);
+    libP2PPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK).finish(__ -> {});
     libP2PPeer.disconnectImmediately(Optional.of(DisconnectReason.REMOTE_FAULT), false);
 
     verify(connection, times(1)).close();

